Cannot read properties of null (reading 'buckets')

```
|14:40:28.195||
|---|---|
|14:40:28.195|  Edits that have been made via the dashboard will be overridden by your local code and config.|
|14:40:28.195||
|14:40:28.195||
|14:40:28.196|? Would you like to continue?|
|14:40:28.196|🤖 Using fallback value in non-interactive context: yes|
|14:40:29.607|🌀 Building list of assets...|
|14:40:29.637|✨ Read 818 files from the assets directory /opt/buildhome/repo/dist|
|14:40:30.567|🌀 Starting asset upload...|
|14:40:31.471||
|14:40:31.472|✘ [ERROR] Cannot read properties of null (reading 'buckets')|
```

Hello, I m trying to build my astro app which works perfectly with pages, in workers instead I get the error:

```
✘ [ERROR] Cannot read properties of null (reading 'buckets')
```

at the step `Deploying to Cloudflare's global network`

which does not make clear of where the problem is.

`wrangler.jsonc`:

```
/**
 * For more details on how to configure Wrangler, refer to:
 * https://developers.cloudflare.com/workers/wrangler/configuration/
 */
{
	"$schema": "node_modules/wrangler/config-schema.json",
	"name": "insertadental-co-uk",
	"account_id": "xxx",
	"main": "./dist/_worker.js/index.js",
	"compatibility_date": "2025-05-31",
	"compatibility_flags": [
		"nodejs_compat",
		"global_fetch_strictly_public"
	],
	"assets": {
		"binding": "ASSETS",
		"directory": "./dist"
	},
	"observability": {
		"enabled": true
	},
	"vars": {
		"PUBLIC_STRIPE_KEY": "xxx",
		"BETTER_AUTH_URL": "xxx",
		"DB_ID": "xxx",
		"CLOUDFLARE_ACCOUNT_ID": "xxx",
		"NODE_VERSION": "22"
	},
	"d1_databases": [
		{
			"binding": "DB",
			"database_name": "inserta-dental",
			"database_id": "xxx"
		}
	],
	"kv_namespaces": [
		{
			"binding": "SESSION",
			"id": "a52dfd465a0a4ae389c5f6958f8e33bd"
		}
	],
	"r2_buckets": [
		{
			"binding": "R2",
			"bucket_name": "inserta-dental"
		}
	]
	/**
	 * Smart Placement
	 * Docs: https://developers.cloudflare.com/workers/configuration/smart-placement/#smart-placement
	 */
	// "placement": { "mode": "smart" },
	/**
	 * Bindings
	 * Bindings allow your Worker to interact with resources on the Cloudflare Developer Platform, including
	 * databases, object storage, AI inference, real-time communication and more.
	 * https://developers.cloudflare.com/workers/runtime-apis/bindings/
	 */
	/**
	 * Environment Variables
	 * https://developers.cloudflare.com/workers/wrangler/configuration/#environment-variables
	 */
	// "vars": { "MY_VARIABLE": "production_value" },
	/**
	 * Note: Use secrets to store sensitive data.
	 * https://developers.cloudflare.com/workers/configuration/secrets/
	 */
	/**
	 * Static Assets
	 * https://developers.cloudflare.com/workers/static-assets/binding/
	 */
	// "assets": { "directory": "./public/", "binding": "ASSETS" },
	/**
	 * Service Bindings (communicate between multiple Workers)
	 * https://developers.cloudflare.com/workers/wrangler/configuration/#service-bindings
	 */
	// "services": [{ "binding": "MY_SERVICE", "service": "my-service" }]
}
```

--- a/.changeset/olive-heads-arrive.md
+++ b/.changeset/olive-heads-arrive.md
@@ -1,0 +1,7 @@
+---
+"wrangler": patch
+---
+
+Fix unclear error when assets upload session returns a `null` response
+
+When deploying assets, if the Cloudflare API returns a `null` response object, Wrangler now provides a clear error message asking users to retry instead of failing with a confusing error.

--- a/packages/wrangler/src/__tests__/deploy/assets.test.ts
+++ b/packages/wrangler/src/__tests__/deploy/assets.test.ts
@@ -1195,5 +1195,37 @@ describe("deploy", () => {
 			});
 			await runWrangler("deploy --dispatch-namespace my-namespace");
 		});
+
+		it("should error when assets upload session response is null", async () => {
+			const assets = [{ filePath: "file-1.txt", content: "Content of file-1" }];
+			writeAssets(assets);
+			writeWranglerConfig({
+				assets: { directory: "assets" },
+			});
+
+			// Mock the AUS endpoint to return a null result
+			msw.use(
+				http.post(
+					`*/accounts/some-account-id/workers/scripts/test-name/assets-upload-session`,
+					() => {
+						return HttpResponse.json(
+							{
+								success: true,
+								errors: [],
+								messages: [],
+								result: null,
+							},
+							{ status: 201 }
+						);
+					}
+				)
+			);
+
+			await expect(
+				runWrangler("deploy")
+			).rejects.toThrowErrorMatchingInlineSnapshot(
+				`[Error: An unexpected response has been received from the Cloudflare API for assets upload. Please try again.]`
+			);
+		});
 	});
 });

--- a/packages/wrangler/src/assets.ts
+++ b/packages/wrangler/src/assets.ts
@@ -71,15 +71,24 @@ export const syncAssets = async (
 
 	// 2. fetch buckets w/ hashes
 	logger.info("🌀 Starting asset upload...");
-	const initializeAssetsResponse = await fetchResult<InitializeAssetsResponse>(
-		complianceConfig,
-		url,
-		{
+	const initializeAssetsResponse =
+		await fetchResult<InitializeAssetsResponse | null>(complianceConfig, url, {
 			headers: { "Content-Type": "application/json" },
 			method: "POST",
 			body: JSON.stringify({ manifest: manifest }),
-		}
-	);
+		});
+
+	// In the past we've seen the endpoint return that incorrectly doesn't contain
+	// a null response (see: https://github.com/cloudflare/workers-sdk/issues/9465).
+	// So just to be extra sure here we check the object and provide a clear error message to the user
+	// if it is falsy.
+	if (!initializeAssetsResponse) {
+		throw new FatalError(
+			"An unexpected response has been received from the Cloudflare API for assets upload. Please try again.",
+			1,
+			{ telemetryMessage: true }
+		);
+	}
 
 	// if nothing to upload, return
 	if (initializeAssetsResponse.buckets.flat().length === 0) {

